src/poetry/puzzle/provider.py

```diff
@@ -363,8 +363,11 @@ def _search_for_file(self, dependency: FileDependency) -> Package:
         return package
 
     @classmethod
-    def get_package_from_file(cls, file_path: Path) -> Package:
+    def get_package_from_file(
+        cls, file_path: Path, source_subdirectory: str | None = None
+    ) -> Package:
         try:
+            PackageInfo._source_subdirectory = source_subdirectory
```

Comment:
Setting the subdirectory as class attribute seems like a crude hack to me. I think that has to be refactored.